// whs/lib/core/RenderingComponent.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.RenderingComponent = undefined;

var _three = require('three');

var THREE = _interopRequireWildcard(_three);

var _ComponentUtils = require('../utils/ComponentUtils');

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function RenderingComponent(target) {
  (0, _ComponentUtils.$defaults)(target, {
    background: {
      color: 0x000000,
      opacity: 1
    },

    shadowmap: {
      enabled: true,
      type: THREE.PCFSoftShadowMap
    },

    renderer: {}
  });

  (0, _ComponentUtils.$define)(target, {
    // TODO
  });

  (0, _ComponentUtils.$extend)(target, {});

  (0, _ComponentUtils.$wrap)(target).onCallConstructor(function (scope) {
    scope.helper = null;
  });

  (0, _ComponentUtils.$wrap)(target).onCallWrap(function (scope) {
    var _native = scope.native,
        _params = scope.params;
  });

  return target;
}

exports.RenderingComponent = RenderingComponent;
