// whs/lib/utils/ComponentUtils.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.$wrap = $wrap;
exports.$extend = $extend;
exports.$define = $define;
exports.$defaults = $defaults;
function $wrap(target) {
  var _proto = target.prototype || target.__proto__;

  return {
    onCallConstructor: function onCallConstructor(callback) {
      _proto.callConstructor = function (old) {
        return function (scope) {
          old(scope);
          callback(scope);
        };
      }(_proto.callConstructor);
    },
    onCallWrap: function onCallWrap(callback) {
      _proto.callWrap = function (old) {
        return function (scope) {
          for (var _len = arguments.length, tags = Array(_len > 1 ? _len - 1 : 0), _key = 1; _key < _len; _key++) {
            tags[_key - 1] = arguments[_key];
          }

          old.apply(undefined, [scope].concat(tags));
          callback.apply(undefined, [scope].concat(tags));
        };
      }(_proto.callWrap);
    },
    onCallAddTo: function onCallAddTo(callback) {
      _proto.callAddTo = function (old) {
        return function (scope) {
          old(scope);
          callback(scope);
        };
      }(_proto.callAddTo);
    },
    onCallCopy: function onCallCopy(callback) {
      _proto.callCopy = function (old) {
        return function (scope) {
          old(scope);
          callback(scope);
        };
      }(_proto.callCopy);
    }
  };
}

function $extend(target, obj) {
  var _proto = target.prototype || target.__proto__;
  Object.assign(_proto, obj);
}

function $define(target, obj) {
  for (var key in obj) {
    var value = obj[key];
    var _proto = target.prototype || target.__proto__;

    value.configurable = true;

    Object.defineProperty(_proto, key, value);
  }
}

function $defaults(target, obj) {
  var defs = target.defaults || {};
  target.defaults = Object.assign(defs, obj);
  target.params = Object.assign(defs, obj);
}
